### PR TITLE
Expand access for ancillary power inputs and total electricity rate/energy to all coil types in UnitarySystem

### LINKED ISSUE #8230: Unitary System Ancillary Electric Power inputs have no effect on results
Issue overview
--------------
I have two IDFs where the only difference between them is that I've set the `Ancillary On-Cycle Electric Power` and `Ancillary Off-Cycle Electric Power` fields to non-zero values for the Unitary System. I expect to see increased energy use but the simulation results are unchanged. The "Unitary System Ancillary Electric Power" output variable also reports zeroes, whereas I expect it to produce 100 W for every timestep.

[zero.idf.txt](https://github.com/NREL/EnergyPlus/files/5120459/zero.idf.txt)
[nonzero.idf.txt](https://github.com/NREL/EnergyPlus/files/5120461/nonzero.idf.txt)

### Details
Some additional details for this issue (if relevant):
 - Windows 10
 - EnergyPlus 9.3

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------
 - Fixes #8230
 - Ancillary On and Ancillary Off power inputs in UnitarySystem were only available for multispeed coil and the new DX cooling coil models. This has been updated to allow these inputs to be used with any coil type. 

   Output:Variable,\*,Unitary System Ancillary Electricity Rate,hourly; !- HVAC Average [W]  
   Output:Variable,\*,Unitary System Cooling Ancillary Electricity Energy,hourly; !- HVAC Sum [J]  
   Output:Variable,\*,Unitary System Heating Ancillary Electricity Energy,hourly; !- HVAC Sum [J]

 - Additionally, the total electricity rate/energy reports had the same issue (only available for multispeed and new coil model) and are now also available for all coil types.

   Output:Variable,\*,Unitary System Electricity Rate,hourly; !- HVAC Average [W]  
   Output:Variable,\*,Unitary System Electricity Energy,hourly; !- HVAC Sum [J]

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 ~~- [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label~~
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 ~~- [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies~~
 - [x] Update documents

Updated defect file to V9.4: 
[8230-nonzero.idf.txt](https://github.com/NREL/EnergyPlus/files/5309313/8230-nonzero.idf.txt)

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/UnitarySystem.cc
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: Moved report variables up so they are available regardless of coil type.

File: src/EnergyPlus/UnitarySystem.cc
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: Delete unused IF

File: src/EnergyPlus/UnitarySystem.cc
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: Delete 2nd IF test for AnyEMSInModel

File: src/EnergyPlus/UnitarySystem.cc
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: Use local variables to store coil electricity use

File: src/EnergyPlus/UnitarySystem.cc
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc:14053]: And expand this type of code to all coil types

File: src/EnergyPlus/UnitarySystem.cc
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc:14069]: Move total power/energy calcs to end and add electricity to the local variable

File: src/EnergyPlus/UnitarySystem.cc
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc:14095]: Water coils don't consume electricity but now include ancillary power/energy calc

File: src/EnergyPlus/UnitarySystem.cc
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: Add supplemental heating coil electricity in total

File: src/EnergyPlus/WaterToAirHeatPump.cc
Comment (rraustad) [inline src/EnergyPlus/WaterToAirHeatPump.cc:1627]: Missing global from water-to-air HP

File: tst/EnergyPlus/unit/UnitarySystem.unit.cc
Comment (rraustad) [inline tst/EnergyPlus/unit/UnitarySystem.unit.cc:856]: Modify existing unit test to include check of ancillary power

**Comment by rraustad** [comment]
Plot of original defect (On=100, Off=100) file for January:

image

Plot of defect file with modified ancillary power inputs, On = 100, Off = 50. The defect file shows 100 W all the time so is rather boring.

image

**Comment by rraustad** [comment]
I am still looking for the reason for diffs in UnitarySystem_WaterCoils_wMultiSpeedFan but for now I show the Unitary System Electricity Rate for this example file for this branch (because this report was not available before for water coils). The answer is identical.

This file has no auxiliary power use so the only energy use is from the fan.

    SetupOutputVariable("Unitary System Electricity Rate", OutputProcessor::Unit::W, unitarySys[sysNum].m_ElecPower,
                                    "System", "Average", unitarySys[sysNum].Name);

    this->m_ElecPower = locFanElecPower + ElecCoolingPower + ElecHeatingPower + SuppHeatingPower + DataHVACGlobals::ElecHeatingCoilPower +
                            this->m_TotalAuxElecPower;

image

File: src/EnergyPlus/UnitarySystem.cc
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc:14226]: Need to remove DataHVACGlobals::ElecHeatingCoilPower in this->m_ElecPower equation.

File: src/EnergyPlus/UnitarySystem.cc
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc:14142]: Sum power across components

**Comment by nrel-bot-2c** [comment]
@rraustad @lgentile it has been 28 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad @lgentile it has been 28 days since this pull request was last updated.

**Comment by rraustad** [comment]
@Myoldmopar there have been a lot of changes lately. I'm not sure what would be easier. To try to merge develop into this branch or use this branch as a template to make the same changes in a fresh branch.

**Comment by Myoldmopar** [comment]
My gut feeling is to just pull in develop to this branch.  If you haven't already started doing it, let me do that tonight or tomorrow morning and push it back up so you are good to continue here.

**Comment by Myoldmopar** [comment]
I went ahead and pulled develop in, wasn't too big of a merge.  Unit test suite passes after the merge so I think everything is fine.

**Comment by rraustad** [comment]
Thank you. This looks like what I remember for this change. This should cause diff's and I'll start back with a review of those to see where this stands.

**Comment by nrel-bot-3** [comment]
@rraustad @lgentile it has been 28 days since this pull request was last updated.

**Comment by Myoldmopar** [comment]
I will pull develop in to resolve the conflict and let CI run once more here to get all green.

File: doc/input-output-reference/src/overview/group-unitary-equipment.tex
Comment (Myoldmopar) [inline doc/input-output-reference/src/overview/group-unitary-equipment.tex:494]: Moving these up, ok.

File: src/EnergyPlus/DataHVACGlobals.hh
Comment (Myoldmopar) [inline src/EnergyPlus/DataHVACGlobals.hh:323]: While having another global isn't great, it fits the existing pattern, so ok.

File: src/EnergyPlus/Furnaces.cc
Comment (Myoldmopar) [inline src/EnergyPlus/Furnaces.cc:309]: Eliminating `using`, nice.

File: src/EnergyPlus/HVACMultiSpeedHeatPump.cc
Comment (Myoldmopar) [inline src/EnergyPlus/HVACMultiSpeedHeatPump.cc:499]: Generalized usage.  Got it.

File: src/EnergyPlus/HVACUnitaryBypassVAV.cc
Comment (Myoldmopar) [inline src/EnergyPlus/HVACUnitaryBypassVAV.cc:4156]: So supplemental heat is set to false here.  I don't quite get why, but maybe I will as I keep reviewing.

File: src/EnergyPlus/HeatingCoils.cc
Comment (Myoldmopar) [inline src/EnergyPlus/HeatingCoils.cc:315]: Maybe this new argument is what allows the calculation to be deferred for supplemental heater operation.

File: src/EnergyPlus/WaterToAirHeatPump.cc
Comment (Myoldmopar) [reply src/EnergyPlus/WaterToAirHeatPump.cc:1627]: OK thanks.

File: tst/EnergyPlus/unit/UnitarySystem.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/UnitarySystem.unit.cc:15673]: OK.

**Review by Myoldmopar — APPROVED** [review]
OK, after looking it all the way over I think I get it.  And the unit test helps.

**Comment by Myoldmopar** [comment]
I pulled in develop locally and resolved the conflict with no problem.  I ran tests and all was good.  I'll push the branch up here now.  You already put some results analysis of the changes here above, but do you have anything to add on the diffs here @rraustad ?

**Comment by rraustad** [comment]
It wouldn't hurt to look at these diff's again before this is merged.

File: src/EnergyPlus/HVACUnitaryBypassVAV.cc
Comment (rraustad) [reply src/EnergyPlus/HVACUnitaryBypassVAV.cc:4156]: CBVAV does not have a supplemental heating coil so the supplemental heating coil flag should be false.

File: src/EnergyPlus/DataHVACGlobals.hh
Comment (rraustad) [reply src/EnergyPlus/DataHVACGlobals.hh:323]: Should these be struct vars instead?

Comment (rraustad) [reply src/EnergyPlus/DataHVACGlobals.hh:323]: No. they can't be struct vars because they are set in the coil models and used in the parent models. So what other methods could eliminate the globals here?

Comment (rraustad) [reply src/EnergyPlus/DataHVACGlobals.hh:323]: A new class I suppose, that is inherited by all who need those vars.

**Comment by rraustad** [comment]
Something isn't right. Looking at the diff's in PTHPVSAS I saw that the only diff was in PTHP2 which has an electric heating coil while the others have fuel and steam. So I thought great, that's the diff I expect. Then I ran that file and looked at the component power. It's way off in winter.

image

And zooming in it appears it's correct in summer. So I suspect that the heating coil is adding energy instead of power.

image

File: src/EnergyPlus/HeatingCoils.cc
Comment (rraustad) [inline src/EnergyPlus/HeatingCoils.cc:2888]: Sure enough, this is after power is converted to energy.

**Comment by rraustad** [comment]
This is after correcting the heating coil code to save power instead of energy.

image

File: src/EnergyPlus/HeatingCoils.cc
Comment (rraustad) [reply src/EnergyPlus/HeatingCoils.cc:315]: Yes, this allows the coil model to save the power number for heating coil and supp coil so when these are added in the parent the correct value is used. Before, the electric supp heater would have overwritten DataHVACGlobals::ElecHeatingCoilPower which is why I needed the new global DataHVACGlobals::SuppHeatingCoilPower.

**Comment by Myoldmopar** [comment]
@rraustad what a find, great job!  OK, so I was already pretty good with these changes, and with your details, I feel completely good about it.  I could still wait on lots of CI results to come in, but I'm just not sure it's needed.  I'll let Mac finish since it already started but I won't waste Linux/Windows cycles on it, as your little follow-up change is obvious.

This will merge later this morning then.  Thanks!

**Comment by rraustad** [comment]
Thanks @Myoldmopar, I am just really glad I did not rely entirely on the diff's as the explanation because I thought the results were correct until I ran the example file and looked closer. Food for thought.

**Comment by Myoldmopar** [comment]
Very excited about this after your fix.  I built and ran locally with the minimal other changes to develop with no problem.  Merging this one in.  Thanks @rraustad !

**Pull request merged**
